FFmpegInterop/Metadata.cpp

```diff
@@ -18,39 +18,140 @@
 
 #include "pch.h"
 #include "Metadata.h"
+#include "FFmpegInteropMSSConfig.h"
 
 using namespace winrt::FFmpegInterop::implementation;
 using namespace winrt;
 using namespace winrt::Windows::Media::Core;
 using namespace winrt::Windows::Storage::Streams;
 using namespace std;
+using namespace wil;
 
 namespace winrt::FFmpegInterop::implementation
 {
-	void PopulateMSSMetadata(_In_ const MediaStreamSource& mss, _In_ const AVDictionary* metadata)
+	com_ptr<IPropertyStore> GetPropertyHandler(_In_ const MediaStreamSource& mss)
 	{
-		// Get the MSS property handler
 		com_ptr<IMFGetService> mediaSource;
 		THROW_IF_FAILED(mss.as<IMFGetService>()->GetService(MF_MEDIASOURCE_SERVICE, __uuidof(mediaSource), mediaSource.put_void()));
 
 		com_ptr<IPropertyStore> propHandler;
 		THROW_IF_FAILED(mediaSource->GetService(MF_PROPERTY_HANDLER_SERVICE, __uuidof(propHandler), propHandler.put_void()));
 
-		// TODO: Populate the property handler with the metadata
+		return propHandler;
 	}
 
-	void SetMSSThumbnail(_In_ const MediaStreamSource& mss, _In_ const AVStream* thumbnailStream)
+	void PopulateMetadata(_In_ const MediaStreamSource& mss, _In_opt_ const AVDictionary* metadata)
 	{
-		// Write the thumbnail to a stream
-		AVPacket_ptr packet{ av_packet_clone(&thumbnailStream->attached_pic) };
+		if (metadata == nullptr)
+		{
+			return;
+		}
+
+		com_ptr<IPropertyStore> propHandler{ GetPropertyHandler(mss) };
+#if DBG
+		com_ptr<IPropertyStoreCapabilities> propHandlerCapabilties{ propHandler.as<IPropertyStoreCapabilities>() };
+#endif // DBG
+
+		const AVDictionaryEntry* tag{ nullptr };
+		while(tag = av_dict_iterate(metadata, tag))
+		{
+			FFMPEG_INTEROP_TRACE("Property: Key = %hs, Value = %hs", tag->key, tag->value);
+
+			// Check if we have a mapping for the property
+			auto iter{ c_metadataPropertiesMap.find(tolower(tag->key)) };
+			if (iter != c_metadataPropertiesMap.end())
+			{
+				const PropertyKey& propKey{ iter->second };
+				unique_prop_variant prop;
+				prop.vt = propKey.type;
+
+				// Convert/parse the property value
+				try
+				{
+					switch (propKey.type)
+					{
+					case VT_LPWSTR:
+						prop.pwszVal = to_cotaskmem_string(tag->value).release();

```

Comment:
does to_cotaskmem_string support char[]?

Reply:
oh you wrote that helper.

Reply:
Yes, to_cotaskmem_string() accepts anything convertible to std::string_view:
```
template <typename T, std::enable_if_t<std::is_convertible_v<T, std::string_view>, int> = 0>
inline wil::unique_cotaskmem_string to_cotaskmem_string(_In_ const T& str)
```

std::string_view can be constructed from a const char*:

[std::basic_string_view<CharT,Traits>::basic_string_view - cppreference.com](https://en.cppreference.com/w/cpp/string/basic_string_view/basic_string_view)
> `constexpr basic_string_view( const CharT* s );`
>
> Constructs a view of the null-terminated character string pointed to by s, not including the terminating null character. The length of the view is determined as if by Traits::length(s). The behavior is undefined if [s, s + Traits::length(s)) is not a valid range. After construction, [data()](https://en.cppreference.com/w/cpp/string/basic_string_view/data) is equal to s, and [size()](https://en.cppreference.com/w/cpp/string/basic_string_view/size) is equal to Traits::length(s).